File: leetcode/arrays-hashing/valid_anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        hm = {}
        for val in s:
            if val not in hm:
                hm[val] = 1
            else:
                hm[val] += 1

        for char in t:
            if char not in hm:
                return False
            else:
                hm[char] -= 1
                if hm[char] == 0:
                    hm.pop(char)

        return len(hm) == 0

# Try it out — press Run
sol = Solution()
print(sol.isAnagram("anagram", "nagaram"))  # True
print(sol.isAnagram("rat", "car"))          # False

File: leetcode/arrays-hashing/valid_anagram.ts
function isAnagram(s: string, t: string): boolean {
  const hm = new Map<string, number>();
  for (const ch of s) {
    hm.set(ch, (hm.get(ch) ?? 0) + 1);
  }
  for (const ch of t) {
    if (!hm.has(ch)) return false;
    const count = hm.get(ch)! - 1;
    if (count === 0) hm.delete(ch);
    else hm.set(ch, count);
  }
  return hm.size === 0;
}

// Try it out — press Run
console.log(isAnagram("anagram", "nagaram")); // true
console.log(isAnagram("rat", "car"));         // false

export {};
